lib/logthrsource/logthrfetcherdrv.c

```diff
@@ -0,0 +1,277 @@
+ * modify it under the terms of the GNU Lesser General Public
+ * License as published by the Free Software Foundation; either
+ * version 2.1 of the License, or (at your option) any later version.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this library; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
+ *
+ * As an additional exemption you are allowed to compile & link against the
+ * OpenSSL libraries as published by the OpenSSL project. See the file
+ * COPYING for details.
+ *
+ */
+
+#include "logthrfetcherdrv.h"
+#include "messages.h"
+
+static EVTTAG *
+evt_tag_driver(LogThreadedFetcherDriver *f)
+{
+  return evt_tag_str("driver", f->super.super.super.id);
+}
+
+static inline void
+_thread_init(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher thread_init()", evt_tag_driver(self));
+  if (self->thread_init)
+    self->thread_init(self);
+}
+
+static inline void
+_thread_deinit(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher thread_deinit()", evt_tag_driver(self));
+  if (self->thread_deinit)
+    self->thread_deinit(self);
+}
+
+static inline gboolean
+_connect(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher connect()", evt_tag_driver(self));
+  if (!self->connect)
+    return TRUE;
+
+  if (!self->connect(self))
+    {
+      msg_debug("Error establishing connection", evt_tag_driver(self));
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static inline void
+_disconnect(LogThreadedFetcherDriver *self)
+{
+  msg_trace("Fetcher disconnect()", evt_tag_driver(self));
+  if (self->disconnect)
+    self->disconnect(self);
+}
+
+static void
+_start_reconnect_timer(LogThreadedFetcherDriver *self)
+{
+  iv_validate_now();
+  self->reconnect_timer.expires  = iv_now;
+  self->reconnect_timer.expires.tv_sec += self->time_reopen;
+  iv_timer_register(&self->reconnect_timer);
+}
+
+static void
+_worker_run(LogThreadedSourceDriver *s)
+{
+  LogThreadedFetcherDriver *self = (LogThreadedFetcherDriver *) s;
+
+  iv_init();
+
+  iv_event_register(&self->wakeup_event);
+  iv_event_register(&self->shutdown_event);
+
+  _thread_init(self);
+  if (_connect(self))
+    iv_task_register(&self->fetch_task);
+  else
+    _start_reconnect_timer(self);
+
+  iv_main();
+
+  _disconnect(self);
+  _thread_deinit(self);
+
+  iv_deinit();
+}
+
+static void
+_worker_request_exit(LogThreadedSourceDriver *s)
+{
+  LogThreadedFetcherDriver *self = (LogThreadedFetcherDriver *) s;
+
+  self->under_termination = TRUE;
+
+  if (self->request_exit)
+    self->request_exit(self);
+
+  iv_event_post(&self->shutdown_event);
+}
```

Comment:
do we need both the virtual method and the event? shouldn't the request exit method post the event? as I understand this is running in the main thread, is that right?

Reply:
No, this is optional, but I find it useful.

Consider a fetcher implementation where a blocking `read()` is used.
In that case, the shutdown event might never be scheduled, so I wanted to give an optional method that can interrupt such a blocking call (a signal in case of `read()` for example).